storage: use RWMutex for {Store,Replica}.mu

petermattis opened the pull request:
Add syncutil.{,RW}Mutex.AssertHeld.

Use reader-locks when reading Store and Replica state. This avoids
contention during read-only code paths.

Remove usage of timedMutex except for Replica.raftMu.

See #13679

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13697)
<!-- Reviewable:end -->

petermattis committed:
storage: use RWMutex for {Store,Replica}.mu

Add syncutil.{,RW}Mutex.AssertHeld.

Use reader-locks when reading Store and Replica state. This avoids
contention during read-only code paths.

Remove usage of timedMutex except for Replica.raftMu.

See #13679

petermattis committed:
storage: add fast-path for lease checking

Use a reader-lock for fast-path lease checking to avoid contention on
Replica.mu.

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 11 of 11 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 915 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdX_V4axDw_5xFeKgWx:-KdX_V4bws36XpVoLH8y:bf7wsud) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/storage/replica.go#L915)):*
> ```Go
> }
> 
> // leaseGoodToGo is a fast-path for lease checks which verifies that an
> ```

this duplicates a bunch of code from redirectOnorAcquireLease. Can it be factored out?

---

*[pkg/util/syncutil/mutex_deadlock.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXZ_nNr0WKnnKdO84T:-KdXZ_nNr0WKnnKdO84U:bdh3f0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1baa854a100474bf3ea25b5c92a6b2b5e47da4cf/pkg/util/syncutil/mutex_deadlock.go#L36)):*
> ```Go
> }
> 
> // AssertHeld is a no-op for deadlock mutexes.
> ```

Hm, how come?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-:-KdX_ZMbK7dZ6w-WUvRA:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 915 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdX_V4axDw_5xFeKgWx:-KdXaFfVxL1RpN08BPrP:bq1uej7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/storage/replica.go#L915)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this duplicates a bunch of code from redirectOnorAcquireLease. Can it be factored out?
</blockquote></details>

I tried initially, but doing so started moving too much code around. It seemed less risky to copy the fast-path checks. @andreimatei or @bdarnell might have some concrete suggestions.

---

*[pkg/util/syncutil/mutex_deadlock.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXZ_nNr0WKnnKdO84T:-KdXaTbudoSRvHY1exww:b-acvsgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1baa854a100474bf3ea25b5c92a6b2b5e47da4cf/pkg/util/syncutil/mutex_deadlock.go#L36)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Hm, how come?
</blockquote></details>

No strong reason, it just didn't seem compelling to implement them given that we usually run with the non-deadlock mutexes which do contain the assertions. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/util/syncutil/mutex_deadlock.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXZ_nNr0WKnnKdO84T:-KdXeqwjGRkzDEWvpII8:b-bah6e8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1baa854a100474bf3ea25b5c92a6b2b5e47da4cf/pkg/util/syncutil/mutex_deadlock.go#L36)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

No strong reason, it just didn't seem compelling to implement them given that we usually run with the non-deadlock mutexes which do contain the assertions. 
</blockquote></details>

Seems easy enough to do =/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 10 of 11 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 734 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXqAyFr-jG6ilihjd1:-KdXqAyFr-jG6ilihjd2:b-r0vh7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/storage/replica.go#L734)):*
> ```Go
> 	ctx context.Context, eng engine.ReadWriter, desc *roachpb.RangeDescriptor,
> ) error {
> 	r.mu.RLock()
> ```

This is an assignment to state in `r.mu`, so it should be a write lock.

---

*[pkg/storage/store.go, line 3094 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXr0xshYJiGyuX9zN_:-KdXr0xshYJiGyuX9zNa:b-6dyaz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/storage/store.go#L3094)):*
> ```Go
> 				// Raft has decided the snapshot shouldn't be applied we would be
> 				// writing the tombstone key incorrectly.
> 				r.mu.RLock()
> ```

This also needs to be a write lock.

---

*[pkg/util/syncutil/mutex_sync.go, line 28 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXpYG_EjOfuxRLTDQ8:-KdXpYG_EjOfuxRLTDQ9:bttfppj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/util/syncutil/mutex_sync.go#L28)):*
> ```Go
> // A Mutex is a mutual exclusion lock.
> type Mutex struct {
> 	sync.Mutex
> ```

Consider un-embedding this since we're no longer using any methods from the `Mutex` type.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-:-KdXrBgQgkYgCd9DCgnw:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 915 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdX_V4axDw_5xFeKgWx:-KdXszU2IIUt8nSqGMvF:b-ggrvct) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/storage/replica.go#L915)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I tried initially, but doing so started moving too much code around. It seemed less risky to copy the fast-path checks. @andreimatei or @bdarnell might have some concrete suggestions.
</blockquote></details>

I don't have any suggestions. This copy doesn't bother me much.

---

*[pkg/storage/replica.go, line 940 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXseRFs95fTIumq_tn:-KdXseRGTTlAdXgyqTtS:b-dzv4a6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/storage/replica.go#L940)):*
> ```Go
> 		}
> 	}
> 	return status, false
> ```

you're returning an empty `LeaseStatus` in the other `false` case above. Can we be consistent, either way?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 915 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdX_V4axDw_5xFeKgWx:-KdXuc8V4YMIbd7PF43U:bvzuvf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/storage/replica.go#L915)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't have any suggestions. This copy doesn't bother me much.
</blockquote></details>

The duplication bothers me a little bit but I don't have any better ideas to dedupe it with `redirectOnOrAcquireLeaderLease`. What about the duplication with `Replica.hasLease`? The code isn't identical (`hasLease` doesn't check for pending transfers), but maybe it should be? If nothing else, we need clearer naming/documentation to indicate when to use `hasLease` and when to use `leaseGoodToGo` and when to use the lower-level `leaseStatus().state == leaseValid`. 

Add a comment that the motivation here is to check the validity of the lease without acquiring `r.mu` in write mode. 

---

*[pkg/storage/replica.go, line 932 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXuqzIu0YlJ3er9Prm:-KdXuqzIu0YlJ3er9Prn:b-wugq7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/storage/replica.go#L932)):*
> ```Go
> 		if status.lease.OwnedBy(r.store.StoreID()) {
> 			// We own the lease...
> 			if repDesc, err := r.getReplicaDescriptorLocked(); err == nil {
> ```

We should probably rename methods like this to `RLocked` to indicate that they only require the write lock (and to avoid the temptation to e.g. cache the result of `getReplicaDescriptorLocked` in a separate field to avoid running the loop repeatedly)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 6 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/replica.go, line 734 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXqAyFr-jG6ilihjd1:-KdY0PaLQnOSS7a-vioC:bdwlxms) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/storage/replica.go#L734)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This is an assignment to state in `r.mu`, so it should be a write lock.
</blockquote></details>

Doh. Fixed.

---

*[pkg/storage/replica.go, line 915 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdX_V4axDw_5xFeKgWx:-KdY0aVwmvtwTO5KruZi:b2p9hqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/storage/replica.go#L915)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The duplication bothers me a little bit but I don't have any better ideas to dedupe it with `redirectOnOrAcquireLeaderLease`. What about the duplication with `Replica.hasLease`? The code isn't identical (`hasLease` doesn't check for pending transfers), but maybe it should be? If nothing else, we need clearer naming/documentation to indicate when to use `hasLease` and when to use `leaseGoodToGo` and when to use the lower-level `leaseStatus().state == leaseValid`. 

Add a comment that the motivation here is to check the validity of the lease without acquiring `r.mu` in write mode. 
</blockquote></details>

The duplication with `hasLease` is subtle. Perhaps I should inline `leaseGoodToGo` into `redirectOnOrAcquireLease`. `leaseGoodToGo` isn't meant be called directly.

---

*[pkg/storage/replica.go, line 932 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXuqzIu0YlJ3er9Prm:-KdY19BY6VXrgq-jHH1w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/storage/replica.go#L932)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We should probably rename methods like this to `RLocked` to indicate that they only require the write lock (and to avoid the temptation to e.g. cache the result of `getReplicaDescriptorLocked` in a separate field to avoid running the loop repeatedly)
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 940 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXseRFs95fTIumq_tn:-KdY1CTy8VOjyRh01dJW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/storage/replica.go#L940)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

you're returning an empty `LeaseStatus` in the other `false` case above. Can we be consistent, either way?
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 3094 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXr0xshYJiGyuX9zN_:-KdY1DbbU75EOWAggiwf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/storage/store.go#L3094)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This also needs to be a write lock.
</blockquote></details>

Done.

---

*[pkg/util/syncutil/mutex_deadlock.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXZ_nNr0WKnnKdO84T:-KdXp90_wDOKNcWH-u1B:b-zhe18t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1baa854a100474bf3ea25b5c92a6b2b5e47da4cf/pkg/util/syncutil/mutex_deadlock.go#L36)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Seems easy enough to do =/
</blockquote></details>

Do we suspect it is going to find anything? When was the last time deadlock 

---

*[pkg/util/syncutil/mutex_sync.go, line 28 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXpYG_EjOfuxRLTDQ8:-KdY1S-ojuEzc7dM822v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/util/syncutil/mutex_sync.go#L28)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Consider un-embedding this since we're no longer using any methods from the `Mutex` type.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 6 of 14 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/util/syncutil/mutex_deadlock.go, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdXZ_nNr0WKnnKdO84T:-KdY4H_F2X9zP6gahWBh:b-7f9gqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1baa854a100474bf3ea25b5c92a6b2b5e47da4cf/pkg/util/syncutil/mutex_deadlock.go#L36)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Do we suspect it is going to find anything? When was the last time deadlock 
</blockquote></details>

Err, when was the last time the deadlock mutex found a problem?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 6 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/replica.go, line 915 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697#-KdX_V4axDw_5xFeKgWx:-Kda3M1AmUtoclPKQRx7:bvtmjk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b0d43b3adbbb966d8a49925b12431619ce85084/pkg/storage/replica.go#L915)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The duplication with `hasLease` is subtle. Perhaps I should inline `leaseGoodToGo` into `redirectOnOrAcquireLease`. `leaseGoodToGo` isn't meant be called directly.
</blockquote></details>

Renamed `hasLease` to `ownsValidLease` and added some commentary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13697)*
<!-- Sent from Reviewable.io -->
